### PR TITLE
feat(ingest): add YamlFileUpdater utility

### PR DESCRIPTION
**Pull request by hsheth2**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/utilities/yaml_sync_utils.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import ruamel.yaml.util
+from ruamel.yaml import YAML
+
+
+@contextlib.contextmanager
+def YamlFileUpdater(file: pathlib.Path) -> Iterator[Any]:
+    yaml = YAML()
+    yaml.preserve_quotes = True  # type: ignore[assignment]
+
+    doc = yaml.load(file)
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/utilities/yaml_sync_utils.py:14]: Why did they make this stateful YAML object the way of doing this lol. I am not a fan

File: metadata-ingestion/tests/unit/utilities/test_yaml_sync_utils.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  - key1: value1
+    key2: value2
+"""
+    )
+    # ind=4, bsi=2
+
+    with YamlFileUpdater(infile) as doc:
+        doc["foo"] = "bar"
+        doc["list_ty"].append("baz")
+        doc["list_ty"][1]["key1.5"] = "val1.5"
```
Comment (asikowitz) [inline metadata-ingestion/tests/unit/utilities/test_yaml_sync_utils.py:28]: Can you test replacing data too? This looks like it's only adding new key-values or list entries

**Review by asikowitz — APPROVED** [review]
Couple questions but LGTM

File: metadata-ingestion/src/datahub/utilities/yaml_sync_utils.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import contextlib
+import pathlib
+from typing import Any, Iterator
+
+import ruamel.yaml.util
+from ruamel.yaml import YAML
```
[the remark this answers is not in the document]
Comment (hsheth2) [reply metadata-ingestion/src/datahub/utilities/yaml_sync_utils.py:6]: pyyaml is the more "standard" option, so we'll likely continue using it in other places. We could migrate, but there's not a strong impetus for it yet.

ruamel is a fork of pyyaml that adds the capability to preserve comments and other formatting during a roundtrip

File: metadata-ingestion/tests/unit/utilities/test_yaml_sync_utils.py (hunk 2 above)
Comment (hsheth2) [reply metadata-ingestion/tests/unit/utilities/test_yaml_sync_utils.py:28]: the real purpose of this test is checking that the comments are retained - the changes themselves are just dummy changes

**Pull request merged**